File: Midterm/Question 3/Cpp/filetrimmer.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <err.h>

void fileManipulation(std::string fileName){
    std::fstream iFile;
    std::fstream oFile;
    int lineCount = 1;

    iFile.open(fileName, std::ios::in);
    if(!iFile.is_open())
        err(1, "!COULD NOT OPEN INPUT FILE!");

    oFile.open("corrected_"+fileName, std::ios::out);
    if(!oFile.is_open())
        err(1, "!COULD NOT OPEN OUTPUT FILE!");

    std::string line;
    while(std::getline(iFile, line))
    {
        if(line.find_first_not_of(" \t\r\n") != std::string::npos)
        {
            line.erase(0, line.find_first_not_of(" \t\r\n"));
            line.erase(line.find_last_not_of(" \t\r\n") + 1);
            oFile << std::to_string(lineCount) + ": " + line + "\n"; 
            lineCount++;
        }
    }
    if(iFile.is_open())
        iFile.close();
    if(oFile.is_open())
        oFile.close();
}

int main()
{
    fileManipulation("CoolText.txt");
}

File: Midterm/Question 3/Clojure/filetrimmer.clj
(require '[clojure.java.io :as io]
         '[clojure.string :as str])

(defn file-manipulation [file-name]
  (let [reader (try (io/reader file-name)
                    (catch java.io.FileNotFoundException e
                      (println "!COULD NOT OPEN INPUT FILE!:" (.getMessage e))
                      (System/exit 1)))]
    (with-open [reader reader]
      (let [writer (try (io/writer (str "corrected_" file-name))
                        (catch java.io.IOException e
                          (println "!COULD NOT OPEN OUTPUT FILE!:" (.getMessage e))
                          (System/exit 1)))]
        (with-open [writer writer]
          (let [lines     (line-seq reader)
                non-blank (filter #(not (str/blank? %)) lines)
                trimmed   (map str/trim non-blank)
                numbered  (map-indexed #(str (+ 1 %1) ": " %2) trimmed)]
            (doseq [line numbered]
              (.write writer (str line "\n")))))))))

(file-manipulation "CoolText.txt")